### PR TITLE
Added BoxResize tests

### PR DESCRIPTION
**Pull request by pepak13**
## Description

Added a unit test to test whether the BoxResize class is working properly. 

Current issues still in of work: 
1) Cannot properly rescale the particle positions. 
2) Using `hoomd.update.BoxResize.linear_volume` for one of the tests is starting with the end box. 

## Motivation and Context

Resolves: 

#700 

## How Has This Been Tested?

Tested changes to single box dimensions, multiple box dimenions, and all box dimensions. 

## Change log

```
Added BoxResize tests. 
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).

### CODE REVIEW

File: hoomd/pytest/test_box_resize.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+from math import isclose
+import numpy as np
+from pytest import fixture
+
+import hoomd
+
+
+class SimulationSetup:
```
Comment (joaander) [inline hoomd/pytest/test_box_resize.py:8]: This would be easier to read if it was written as one test_ method using `pytest.mark.parameterize` to give it different box1 and box2 inputs. You could also parameterize with a flag to test either `BoxResize` or `BoxResize.linear_variant`.

File: hoomd/pytest/test_box_resize.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        trigger = hoomd.trigger.On(self.t_start)
+        box_resize = hoomd.update.BoxResize(
+            box1=self.box1, box2=self.box2,
+            variant=variant, trigger=trigger, scale_particles=scale_particles)
+        sim.operations.updaters.append(box_resize)
+        sim.run(self.t_start + self.t_ramp + 1)
+        return box_resize, sim.state.snapshot
+
+    def variant_linear(self, device, scale_particles=True):
+        sim = hoomd.Simulation(device)
```
Comment (joaander) [inline hoomd/pytest/test_box_resize.py:45]: Please use the `simulation_factory` fixture to create `Simulation` objects from snapshots.

File: hoomd/pytest/test_box_resize.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        return box_resize, sim.state.snapshot
+
+
+def test_resize_Lx(device):
+    box1_dim = np.array([1., 2., 3., 1., 2., 3.])
+    box2_dim = np.array([0.5, 2., 3., 1., 2., 3.])
+    ss = SimulationSetup(box1_dim=box1_dim, box2_dim=box2_dim)
+    box_resize, snap = ss.variant_ramp(device)
+
+    assert np.all(box_resize.get_box(0).matrix == ss.box1.matrix)
```
Comment (joaander) [inline hoomd/pytest/test_box_resize.py:64]: You should be able to check for equality between two box objects directly.

File: hoomd/pytest/test_box_resize.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    assert np.all(box_resize.get_box(ss.t_start + ss.t_ramp).matrix == ss.box2.matrix)
+    # assert np.all(np.isclose(snap.particles.position[:], ss.points2))
+
+    box_resize, snap = ss.variant_linear(device)
+
+    assert np.all(box_resize.get_box(0).matrix == ss.box1.matrix)
+    assert np.all(box_resize.get_box(ss.t_start + ss.t_ramp).matrix == ss.box2.matrix)
+    # assert np.all(np.isclose(snap.particles.position[:], ss.points2))
```
Comment (joaander) [inline hoomd/pytest/test_box_resize.py:72]: Use `numpy.testing.assert_allclose`

File: hoomd/pytest/test_box_resize.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+import numpy as np
+import numpy.testing as npt
+import pytest
+import hoomd
+
+
+@pytest.fixture(scope="module")
+def fractional_coordinates(n=10):
+    """
+    TODO: Does `numpy_random_seed()` in conftest.py run for this function?
```
Comment (joaander) [inline hoomd/pytest/test_box_resize.py:11]: I don't know. That fixture is set with `@pytest.fixture(scope='function', autouse=True)` so it will be called before every function, but I don't know how that works with a module scope fixture.

You could make sure that it works in the correct order by making this a function scope fixture and include numpy_random_seed as an argument to ensure that that fixture is called first.

**Review by joaander — COMMENTED** [review]
Looking good! @b-butler do you have any comments?

Do you have any additional tests or modifications you plan to make?

File: hoomd/pytest/test_box_resize.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    assert box_resize.get_box(0) == sys1[0]
+    assert box_resize.get_box(variant.t_start*3) == sys2[0]
+    assert sim.state.box == sys2[0]
+    if scale_particles:
+        npt.assert_allclose(sys2[1], sim.state.snapshot.particles.position)
+    else:
+        npt.assert_allclose(sys1[1], sim.state.snapshot.particles.position)
+    if variant == _variants[2]:
+       assert box_resize.get_box(variant.t_start*3+1) == sys1[0]
```
Comment (b-butler) [inline hoomd/pytest/test_box_resize.py:108]: I think a comment explicitly mentioning this is checking that we are correctly handling the `Cycle` variant is needed. Also, this should be in the `test_get_box` test I mention above.

File: hoomd/pytest/test_box_resize.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+
+_scale_particles = [True,False]
+
+
+@pytest.fixture(scope='function', params=_scale_particles)
+def scale_particles(request):
+    return request.param
+
+
```
Comment (b-butler) [inline hoomd/pytest/test_box_resize.py:86]: For this fixture we could just use `pytest.mark.parameterize` since it is only used for this function and only for the simple values `True` and `False`.

**Review by b-butler — CHANGES_REQUESTED** [review]
It would be helpful to check intermediate box and particle positions as @joaander mentions. It may be worth creating a function given an initial box and particles that when passed a variant value can determine the new box and particles that are expected. An easier if less accurate approach would be to check that the box sizes are exactly the initial or final conditions at the correct time steps or strictly in between for other time steps in this approach positions would just use an equality or non equal check.

Also some more modularity in tests would be nice. Ideally a test should test one thing.

That being said great work and thanks for the effort thus far.

**Comment by pepak13** [comment]
> Looking good! @b-butler do you have any comments?
> 
> Do you have any additional tests or modifications you plan to make?

File: hoomd/pytest/test_box_resize.py (hunk 5 above)
Comment (pepak13) [reply hoomd/pytest/test_box_resize.py:11]: I want to note that the linear_volume box resize test is still failing for one of the systems. In addition, the other variant tests pass when the xy for sys21 is set to 0.9 (It fails when it is 2). I still do not know why this is happening.

Comment (vyasr) [reply hoomd/pytest/test_box_resize.py:11]: @joaander for any test requiring randomness, it might make sense to instead define a fixture
```
def rng():
    return numpy.random.default_rng(42)
```

This has the downside of adding to the signature of each test, but it has the upside of being self-contained and not having to worry about global numpy random state.

Comment (vyasr) [reply hoomd/pytest/test_box_resize.py:11]: There is `pytest-rng`, but that's still using the legacy numpy random number API (and this is a trivial fixture to add anyway).

Comment (joaander) [reply hoomd/pytest/test_box_resize.py:11]: > but it has the upside of being self-contained and not having to worry about global numpy random state.
Yes, this was my intention with the autouse fixture. We could use an explicit fixture instead.

**Comment by klywang** [comment]
Some updates:
* Separated tests for `get_box` and `BoxResize` on the simulation state.
* Added assertions for intermediate particle positions and box size.
* Test 2D and 3D box resizing.
* Added tests to ensure `variant` and  `trigger`.
* Restructured tests with classes.
    - I found it easier to determine which test failed with the classes. Please let me know if I should change the structure back to functions only.

File: sphinx-doc/credits.rst [hunk 8]
```diff
@@ -95,3 +95,868 @@ The following people have contributed to the to HOOMD-blue:
 * Wouter Ellenbroek, Eindhoven University of Technology
 * Yuan Zhou, University of Michigan
 * Åsmund Ervik, SINTEF
+
+=======
```
Comment (b-butler) [inline sphinx-doc/credits.rst:99]: Why are these changes here? Are they just hold outs from a removal in master?

File: hoomd/pytest/test_box_resize.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    def test_get_box(self, device, simulation_factory,
+                     get_snapshot, sys, box_resize):
+        sys1, make_sys_halfway, sys2 = sys
+        sys_halfway = make_sys_halfway(self._power)
+
+        sim = hoomd.Simulation(device)
+        sim.create_state_from_snapshot(get_snapshot())
+
+        sim.operations.updaters.append(box_resize)
+        sim.run(_t_start + _t_ramp)
```
Comment (b-butler) [inline hoomd/pytest/test_box_resize.py:125]: Is there any need to run? Could we just use `sim.run(0)`?

File: hoomd/pytest/test_box_resize.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        assert sim.state.box == sys_halfway[0]
+        npt.assert_allclose(sim.state.snapshot.particles.position, sys1[1])
+
+        # Finish run
+        sim.run(_t_mid)
+        assert sim.state.box == sys2[0]
+        npt.assert_allclose(sim.state.snapshot.particles.position, sys1[1])
+
+
+class TestLinearVolume(TestBoxResize):
```
Comment (b-butler) [inline hoomd/pytest/test_box_resize.py:193]: I think the idea here would be to test the factory method `hoomd.update.BoxResize.linear_volume()`. Basically we just want to test that the volume changes linearly over the box change.

**Review by b-butler — CHANGES_REQUESTED** [review]
I left a few comments to address.

File: sphinx-doc/credits.rst (hunk 8 above)
Comment (joaander) [reply sphinx-doc/credits.rst:99]: This looks like a merge conflict that was not resolved: it starts with `=======`

Comment (klywang) [reply sphinx-doc/credits.rst:99]: It is a merge conflict that I failed to resolve correctly. Do I remove everything from line 3 to "====="  on line 99?

File: hoomd/pytest/test_box_resize.py (hunk 10 above)
Comment (klywang) [reply hoomd/pytest/test_box_resize.py:193]: I tried to test if box volume changed linearly with code like this:
```python
    def test_linear_volume(self, device, simulation_factory
        sim = hoomd.Simulation(device)
        sim.create_state_from_snapshot(get_snapshot())
        sim.operations.updaters.append(box_resize)
        sim.run(0)

        slope = (sys2[0].volume - sys1[0].volume)/_t_ramp
        for ts in range(_t_ramp + 1):
            estimated_volume = sys1[0].volume + ts*slope
            assert box_resize.get_box(ts + _t_start).volume == estimated_volume
```
and the test failed.

When I plot box volume for a 3D box, I get this:
linear_volume_3d



and when I plot box area for a 2D, I get this:
linear_volume2d



Am I testing/interpreting this correctly?

Comment (klywang) [reply hoomd/pytest/test_box_resize.py:193]: Does that mean we don't need to test things like `get_box` and `trigger` for the factory method? And what does the `variant` control? If I understand correctly, in order for `hoomd.update.BoxResize.linear_volume` is supposed to scale volume linearly, it uses it's own calculations to determine the appropriate coefficients for the `hoomd.variant.Power`.

File: sphinx-doc/credits.rst (hunk 8 above)
Comment (joaander) [reply sphinx-doc/credits.rst]: I would recommend just replacing the credits.rst file with the one from master. Then add your names to it.

File: hoomd/pytest/test_box_resize.py (hunk 10 above)
Comment (joaander) [reply hoomd/pytest/test_box_resize.py:193]: @klywang Are you test boxes cubic?

Comment (klywang) [reply hoomd/pytest/test_box_resize.py:193]: @joaander These are not.

Comment (joaander) [reply hoomd/pytest/test_box_resize.py:193]: Does the plot look right for cubic boxes? I'm thinking that the the power variant math @b-butler wrote may have am implicit assumption that the box is cubic.

Comment (b-butler) [reply hoomd/pytest/test_box_resize.py:193]: I think it does. I didn't realize it when writing it though.

Comment (b-butler) [reply hoomd/pytest/test_box_resize.py:193]: To get linear volume change we need to either change the way `BoxResize` works or have an internal flag that changes the behavior to allow for linear volume resizing.

Comment (klywang) [reply hoomd/pytest/test_box_resize.py:193]: Here are the results for `BoxResize.linear_volume` for 3D cubes:
box_resize3D

And 2D squares:
box_resize2D

**Comment by joaander** [comment]
Discussed offline: boxes formed from `s*box1 + (1-s)*box2` don't follow a linear volume scaling when s follows a 1/3 power. Please remove the linear_volume factory method and any tests for it.

**Review by b-butler — APPROVED** [review]
LGTM

**Review by joaander — APPROVED** [review]
Thanks!
